Blog: Add a guide about critical CSS tools (#898)
GoogleChrome/web.dev: src/site/content/en/blog/tools-for-extracting-critical-css/index.md

description: |
  Learn how to improve render times with critical CSS technique and how to choose the best tool for your project.
tags:
  - post # post is a required tag for the article to show up in the blog.
  - performance
  - css
---

The browser must download and parse CSS files before it can show the page, which makes CSS a render-blocking resource. If CSS files are big, or network conditions are poor, requests for CSS files can significantly increase the time it takes for a web page to render.

{% Aside 'key-term' %}
Critical CSS is a technique that extracts the CSS for above-the-fold content in order to get a meaningful render as fast as possible.
{% endAside %}

<figure class="w-figure  w-figure--center">
  <img class="" src="./above-the-fold.png" alt="An illustration of a laptop and a mobile device with web pages overflowing the edges of screens" style="max-width: 600px;">
</figure>

{% Aside 'note' %}
Above-the-fold is all the content a viewer sees on page load, before scrolling. There is no universally defined pixel height of what is considered above the fold content since there is a myriad of devices and screen sizes.
{% endAside %}

> Inlining extracted styles in the `<head>` of the HTML document eliminates the need to make an additional request to fetch these styles. The remainder of the CSS can be loaded asynchronously.

Reviewer: Reading the current text, i think we could tweak our description of how roundtrips to first render should be minimized. We're saying here that you should keep your HTML and your inline CSS under 14KB. Some of our prior guidance has said it's the above-the-old content that should fit under 14KB.

"…the server can send up to 10 TCP packets on a new connection (~14KB) in first roundtrip, and then it must wait for the client to acknowledge this data before it can grow its congestion window and proceed to deliver more data.

Due to this TCP behavior, it is important to optimize your content to minimize the number of roundtrips required to deliver the necessary data to perform the first render of the page. Ideally, the ATF (above the fold) content should fit under 14KB — this allows the browser to paint the page after just one roundtrip…"

https://developers.google.com/speed/docs/insights/mobile#delivering-the-sub-one-second-rendering-experience

---
title: Tools for extracting critical CSS
subhead: Learn how to improve render times with critical CSS technique.
authors:
  - mihajlija
date: 2019-05-07
hero: hero.jpg
alt: A description of the hero image for screen reader users.
description: |
  Learn how to improve render times with critical CSS technique and how to choose the best tool for your project.
tags:
  - post # post is a required tag for the article to show up in the blog.
  - performance
  - css
---

The browser must download and parse CSS files before it can show the page, which makes CSS a render-blocking resource. If CSS files are big, or network conditions are poor, requests for CSS files can significantly increase the time it takes for a web page to render.

{% Aside 'key-term' %}

> Critical CSS is a technique that extracts the CSS for above-the-fold content in order to get a meaningful render as fast as possible.

Reviewer 2: ```suggestion
Critical CSS is a technique that extracts the CSS for above-the-fold content in order to render content to the user as fast as possible.
```

{% endAside %}

<figure class="w-figure  w-figure--center">

>   <img class="" src="./above-the-fold.png" alt="An illustration of a laptop and a mobile device with web pages overflowing the edges of screens" style="max-width: 600px;">

Reviewer 3: ```suggestion
  <img class="" src="./above-the-fold.png" alt="An illustration of a laptop and a mobile device with web pages overflowing the edges of screens" style="max-width: 600px; width: 100%;">
```

I think this is needed to properly scale it on mobile.

</figure>

{% Aside 'note' %}
Above-the-fold is all the content a viewer sees on page load, before scrolling. There is no universally defined pixel height of what is considered above the fold content since there is a myriad of devices and screen sizes.
{% endAside %}

Inlining extracted styles in the `<head>` of the HTML document eliminates the need to make an additional request to fetch these styles. The remainder of the CSS can be loaded asynchronously.

<figure class="w-figure  w-figure--center">

>     <img class="w-screenshot" src="./inline-critical-css.png" alt="HTML file with critical CSS inlined in the head">

Reviewer 3: Is it possible to recreate this screenshot with the font size boosted?

    <figcaption class="w-figcaption">
    Inlined critical CSS
    </figcaption>
</figure>

Improving render times can make a huge difference in [perceived performance](https://developers.google.com/web/fundamentals/performance/rail#ux), especially under poor network conditions. On mobile networks, high latency is an issue regardless of bandwidth.

<figure class="w-figure  w-figure--center">
  <img class="w-screenshot" src="./filmstrip-comparison.png" alt="Filmstrip view comparison of loading a page with render-blocking CSS (top) and the same page with inlined critical CSS (bottom) on a 3G connection. Top filmstrip shows six blank frames before finally displaying content. Bottom filmstrip displays meaningful content in the first frame.">
  <figcaption class="w-figcaption">
    Comparison of loading a page with render-blocking CSS (top) and the same page with inlined critical CSS (bottom) on a 3G connection
  </figcaption>
</figure>

If you have poor [First Contentful Paint](https://web.dev/first-contentful-paint/) (FCP) and see "Eliminate render-blocking resource" opportunity in Lighthouse audits it’s a good idea to give critical CSS a go.

> <img class="w-screenshot" src="./lighthouse-audit.png" alt="Lighthouse audit with 'Eliminate render-blocking resource' or 'Defer unused CSS' opportunities">

Reviewer 3: I was wondering if you have a higher res version of this image? The text seems a bit fuzzy. We'll run the image through compression before it goes live so you don't have to worry much about compressing beforehand. So long as images are sized correctly, we should be able to handle things after that.

{% Aside 'gotchas' %}
Keep in mind that if you inline a large amount of CSS, it delays the transmission of the rest of the HTML document. If everything is prioritized then nothing is. Inlining also has some downsides in that it prevents the browser from caching the CSS for reuse on subsequent page loads, so it’s best to use it sparingly.
{% endAside %}

<p id="14KB">To minimize the number of roundtrips to first render, aim to keep above-the-fold content under <strong>14 KB</strong> (compressed).</p>

{% Aside 'note' %}

> New [TCP](https://hpbn.co/building-blocks-of-tcp/) connections cannot immediately use the full available bandwidth between the client and the server, they all go through [slow-start](https://hpbn.co/building-blocks-of-tcp/#slow-start) to estimate the connection capacity. In this process, the server starts the transfer with a small amount of data and if it reaches the client in perfect condition, doubles the amount in the next roundtrip. For most servers, 10 packets or approximately 14 KB is the maximum that can be transferred in the first roundtrip.

Reviewer 2: ```suggestion
New [TCP](https://hpbn.co/building-blocks-of-tcp/) connections cannot immediately use the full available bandwidth between the client and the server, they all go through [slow-start](https://hpbn.co/building-blocks-of-tcp/#slow-start) to avoid overloading the connection with more data than it can carry. In this process, the server starts the transfer with a small amount of data and if it reaches the client in perfect condition, doubles the amount in the next roundtrip. For most servers, 10 packets or approximately 14 KB is the maximum that can be transferred in the first roundtrip.
```

{% endAside %}

The performance impact you can achieve with this technique depends on the type of your website. Generally speaking, the more CSS a site has, the greater the possible impact of inlined CSS.

## Overview of tools

> Determining the critical CSS for a page is rather complex. It requires analysis of the entire DOM to determine the styles that are applied to each element in the viewport. Doing this manually would be a tedious process, but there are a number of great tools that do this automatically. 

Reviewer 2: ```suggestion
There are a number of great tools that can automatically determine the critical CSS for a page. This is good news because doing this manually would be a tedious process. It requires analysis of the entire DOM to determine the styles that are applied to each element in the viewport.
```

### Critical

[Critical](https://github.com/addyosmani/critical) extracts, minifies and inlines above-the-fold CSS and is available as [npm module](https://www.npmjs.com/package/critical). It can be used with Gulp (directly) or with Grunt (as a [plugin](grunt-critical)) and there’s a [webpack plugin](html-critical-webpack-plugin) too.

It’s a simple tool that takes a lot of thinking out of the process. You don’t even have to specify the stylesheets, Critical automatically detects them. It also supports extracting critical CSS for multiple screen resolutions.

### criticalCSS

[CriticalCSS](https://github.com/filamentgroup/criticalCSS) is another [npm module](https://www.npmjs.com/package/criticalcss) that extracts above-the-fold CSS. It is also available as a CLI. 

It doesn’t have options to inline and minify critical CSS, but it does let you force-include rules that don't actually belong in critical CSS and gives you more granular control over including `@font-face` declarations.

### Penthouse

> [Penthouse](https://github.com/pocketjoso/penthouse) is a good choice if your site or app has a large number of styles or styles which are being dynamically injected into the DOM (common in Angular apps). It is a more robust tool that uses [Puppeteer](https://github.com/GoogleChrome/puppeteer) under the hood. It features an [online hosted version](https://jonassebastianohlsson.com/criticalpathcssgenerator/) too. 

Reviewer 2: ```suggestion
[Penthouse](https://github.com/pocketjoso/penthouse) is a good choice if your site or app has a large number of styles or styles which are being dynamically injected into the DOM (common in Angular apps). It uses [Puppeteer](https://github.com/GoogleChrome/puppeteer) under the hood and even features an [online hosted version](https://jonassebastianohlsson.com/criticalpathcssgenerator/) too. 
```
